presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CombineApproxPercentileFunctions.java

```diff
@@ -0,0 +1,296 @@
+                    newAggCallArguments);
+
+            AggregationNode.Aggregation newAggregation =
+                    new AggregationNode.Aggregation(
+                            approxPercentileCall,
+                            aggregationBeforeMerge.getFilter(),
+                            aggregationBeforeMerge.getOrderBy(),
+                            aggregationBeforeMerge.isDistinct(),
+                            aggregationBeforeMerge.getMask());
+            // Get variable reference to the new aggregation
+            VariableReferenceExpression newVariableReference = context.getVariableAllocator().newVariable(newAggregation.getCall());
+            aggregations.put(newVariableReference, newAggregation);
+
+            variableToArrayIndex.forEach((variable, index) -> {
+                RowExpression elementAtExpression = call(
+                        functionAndTypeManager,
+                        "element_at",
+                        colType,
+                        ImmutableList.of(newVariableReference, constant((long) index, BIGINT)));
+                outputProjectAssignments.put(variable, elementAtExpression);
+                combinedAggVariableReference.add(variable);
+            });
+        }
+
+        if (!hasChanged) {
+            return Result.empty();
+        }
+
+        // Add aggregations which are not combined to the new aggregation node.
+        aggregation.getAggregations().forEach((key, value) -> {
+            if (!combinedAggVariableReference.contains(key)) {
+                aggregations.put(key, value);
+            }
+        });
+
+        // Add output of the old aggregations which are not changed in the rewrite to the parent projection node.
+        aggregation.getOutputVariables().forEach(variable -> {
+            if (!combinedAggVariableReference.contains(variable)) {
+                outputProjectAssignments.put(variable, variable);
+            }
+        });
+
+        aggregation.getSource().getOutputVariables().forEach(variable -> sourceProjectAssignments.put(variable, variable));
+
+        return Result.ofPlanNode(
+                new ProjectNode(context.getIdAllocator().getNextId(),
+                        new AggregationNode(
+                                aggregation.getSourceLocation(),
+                                context.getIdAllocator().getNextId(),
+                                new ProjectNode(context.getIdAllocator().getNextId(),
+                                        aggregation.getSource(), sourceProjectAssignments.build()),
+                                aggregations.build(),
+                                aggregation.getGroupingSets(),
+                                ImmutableList.of(),
+                                aggregation.getStep(),
+                                aggregation.getHashVariable(),
+                                aggregation.getGroupIdVariable()),
+                        outputProjectAssignments.build()));
+    }
+}
```

Comment:
Hmm - looks like code has gotten a lot longer now :) can we simplify it. My suggestion of building the map will make the code lot smaller:

apprx_pcts = aggregation.getAggregations().stream().filter(x -> isApproxPct(x));

Then iterate on that to split on arg1:
mabuilder apmap_buider
apprx_pcts.stream().map(x -> apmapbuider.put(x.getcall().getargs().get(1), x.getcall());

Now you can simply work on it.

If that's not possible, we can look at this code again

Reply:
Simplify the steps and the logic as suggested.